### PR TITLE
[mtsource] don't kill pods with placed vpods

### PR DESCRIPTION
**Pull request by lionelvillard**
## Proposed Changes

- Limit scaling down the adapter to the last pod with placed "vpods" 
- Temporarily disable cancelling single-tenant adapters. See https://github.com/knative-sandbox/eventing-kafka/issues/382
- 

**Release Note**

```release-note

```

**Docs**

source e2e tests are passing ... locally. Let's see if Prow is happy.

### CODE REVIEW

**Comment by matzew** [comment]
I still see a failure on the MT source job

**Comment by lionelvillard** [comment]
I know :-) Not an easy fix. There are at least 2 other issues I'm currently working on.

File: pkg/common/scheduler/statefulset/snapshot.go [hunk 1]
```diff
@@ -22,6 +22,10 @@ import "go.uber.org/zap"
 type Snapshot struct {
 	// free tracks the free capacity of each pod.
 	free map[string]int32
+
+	// lastOrdinal is the ordinal index corresponding to the last statefulset replica
+	// with placed vpods.
+	lastOrdinal int32
```
Comment (steven0711dong) [inline pkg/common/scheduler/statefulset/snapshot.go:28]: Is the adding of lastOrdinal only for snapshot test?

File: pkg/common/scheduler/statefulset/autoscaler.go [hunk 2]
```diff
@@ -72,13 +73,21 @@ func (s *StatefulSetScheduler) autoscale(ctx context.Context) {
...
-				s.logger.Infow("updating adapter replicas", zap.Int32("replicas", scale.Spec.Replicas))
+				// Make sure not to scale down past the last pod with placed vpods
+				if new < snapshot.lastOrdinal+1 {
+					new = snapshot.lastOrdinal + 1
+				}
 
-				_, err = s.statefulSetClient.UpdateScale(ctx, s.statefulSetName, scale, metav1.UpdateOptions{})
-				if err != nil {
-					s.logger.Errorw("updating scale subresource failed", zap.Error(err))
+				if new != scale.Spec.Replicas {
```
Comment (steven0711dong) [inline pkg/common/scheduler/statefulset/autoscaler.go:83]: Before this change, we always update the statefulset scale, even when new calculated replica number equals the current replicas. What's the drawback of this always update? Why is this update needed?

**Review by steven0711dong — APPROVED** [review]
/lgtm

File: pkg/common/scheduler/statefulset/snapshot.go (hunk 1 above)
Comment (lionelvillard) [reply pkg/common/scheduler/statefulset/snapshot.go:28]: It's is used below, to make sure no pods with placed vpods are killed.

File: pkg/common/scheduler/statefulset/autoscaler.go (hunk 2 above)
Comment (lionelvillard) [reply pkg/common/scheduler/statefulset/autoscaler.go:83]: see comment above.

**Pull request merged**